gateway/src/resty/concurrent/timer_task.lua

```diff
@@ -0,0 +1,94 @@
+local uuid = require 'resty.jit-uuid'
+local setmetatable = setmetatable
+local getmetatable = getmetatable
+local newproxy = newproxy
+local unpack = unpack
+
+local _M = {}
+
+local default_interval_seconds = 60
+
+_M.active_tasks = {}
+
+function _M.register_task(id)
+  _M.active_tasks[id] = true
+end
+
+function _M.unregister_task(id)
+  _M.active_tasks[id] = nil
+end
+
+function _M.task_is_active(id)
+  return _M.active_tasks[id] or false
+end
+
+local function generate_id()
+  return uuid.generate_v4()
+end
+
+local function mt(id)
+  -- Using 'newproxy' is needed to be able to define __gc().
+  -- With __gc we can make sure that the task will stop scheduling more work
+  -- after it has been garbage collected.
+  local proxy = newproxy(true)
+  local res_mt = getmetatable(proxy)
+  res_mt.__gc = function() _M.unregister_task(id) end
+  res_mt.__index = _M
+  return res_mt
+end
+
+--- Initialize a TimerTask.
+-- @tparam function task The function to be run periodically
+-- @tparam[opt] table opts
+-- @tfield ?table args Arguments to the function
+-- @tfield ?number interval Interval in seconds (defaults to 60)
+function _M.new(task, opts)
+  local options = opts or {}
+
+  local id = generate_id()
+
+  local self = setmetatable({}, mt(id))
+  self.task = task
+  self.args = options.args
+  self.interval = options.interval or default_interval_seconds
+  self.id = id
+
+  _M.register_task(id)
+
+  return self
+end
+
+local run_periodic, schedule_next, timer_execute
```

Comment:
That is very weird style. Why not just define each function as local?

Reply:
I see. They all depend on each other so they need to be defined this way.

Reply:
Exactly.